src/MongoDB/TopologyDescription.c

```diff
@@ -0,0 +1,207 @@
+{
+	php_phongo_topologydescription_t* intern;
+	mongoc_server_description_t**     sds;
+	size_t                            i, n = 0;
+
+	intern = Z_TOPOLOGYDESCRIPTION_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	sds = mongoc_topology_description_get_servers(intern->topology_description, &n);
+
+	array_init_size(return_value, n);
+
+	for (i = 0; i < n; i++) {
+		zval obj;
+
+		phongo_server_init(&obj, getThis(), mongoc_server_description_id(sds[i]));
+		add_next_index_zval(return_value, &obj);
+	}
+
+	mongoc_server_descriptions_destroy_all(sds, n);
+} /* }}} */
+
+/* {{{ proto string MongoDB\Driver\TopologyDescription::getType()
+    Returns the topology type */
+static PHP_METHOD(TopologyDescription, getType)
+{
+	php_phongo_topologydescription_t* intern = Z_TOPOLOGYDESCRIPTION_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	RETVAL_STRING(mongoc_topology_description_type(intern->topology_description));
+} /* }}} */
+
+/* {{{ MongoDB\Driver\TopologyDescription function entries */
+ZEND_BEGIN_ARG_INFO_EX(ai_TopologyDescription_void, 0, 0, 0)
+ZEND_END_ARG_INFO()
+
+static zend_function_entry php_phongo_topologydescription_me[] = {
+	/* clang-format off */
+	PHP_ME(TopologyDescription, getServers, ai_TopologyDescription_void, ZEND_ACC_PUBLIC | ZEND_ACC_FINAL)
+	PHP_ME(TopologyDescription, getType, ai_TopologyDescription_void, ZEND_ACC_PUBLIC | ZEND_ACC_FINAL)
+	ZEND_NAMED_ME(__construct, PHP_FN(MongoDB_disabled___construct), ai_TopologyDescription_void, ZEND_ACC_PRIVATE | ZEND_ACC_FINAL)
+	PHP_FE_END
+	/* clang-format on */
+};
+/* }}} */
+
+/* {{{ MongoDB\Driver\TopologyDescription object handlers */
+static zend_object_handlers php_phongo_handler_topologydescription;
+
+static void php_phongo_topologydescription_free_object(zend_object* object) /* {{{ */
+{
+	php_phongo_topologydescription_t* intern = Z_OBJ_TOPOLOGYDESCRIPTION(object);
+
+	zend_object_std_dtor(&intern->std);
+
+	if (intern->properties) {
+		zend_hash_destroy(intern->properties);
+		FREE_HASHTABLE(intern->properties);
+	}
+
+	if (intern->topology_description) {
+		mongoc_topology_description_destroy(intern->topology_description);
+	}
+}
+
+static zend_object* php_phongo_topologydescription_create_object(zend_class_entry* class_type) /* {{{ */
+{
+	php_phongo_topologydescription_t* intern = NULL;
+
+	intern = PHONGO_ALLOC_OBJECT_T(php_phongo_topologydescription_t, class_type);
+
+	zend_object_std_init(&intern->std, class_type);
+	object_properties_init(&intern->std, class_type);
+
+	intern->std.handlers = &php_phongo_handler_topologydescription;
+
+	return &intern->std;
+} /* }}} */
+
+HashTable* php_phongo_topologydescription_get_properties_hash(phongo_compat_object_handler_type* object, bool is_debug) /* {{{ */
+{
+	php_phongo_topologydescription_t* intern = NULL;
+	HashTable*                        props;
+
+	intern = Z_OBJ_TOPOLOGYDESCRIPTION(PHONGO_COMPAT_GET_OBJ(object));
+
+	PHONGO_GET_PROPERTY_HASH_INIT_PROPS(is_debug, intern, props, 2);
```

Comment:
In other places where `PHONGO_GET_PROPERTY_HASH_INIT_PROPS` is called, a subsequent line returns early if `intern` is uninitialized. Depending on the struct definition, those lines either consult a pointer that should be non-zero or a boolean (in cases where the `intern` is so simple it doesn't contain a pointer). That's missing here, and it looks like it was also missed in the PR for ServerDescription.c.

You can address this here, but I'd suggest a separate PR to go back and revise ServerDescription.c.

We don't test for this, but the early return is a sanity check to avoid making further access to `intern` if it happens to be uninitialized (e.g. only zero-allocated).

Reply:
https://jira.mongodb.org/browse/PHPC-1934